### PR TITLE
Preflight - global handlers

### PR DESCRIPTION
**Pull request by tomaj**

### CODE REVIEW

File: src/ApiDecider.php [hunk 1]
```diff
@@ -34,6 +41,13 @@ public function getApiHandler($method, $version, $package, $apiAction = '')
                 $handler['handler']->setEndpointIdentifier($endpointIdentifier);
                 return $handler;
             }
+            if ($method == 'OPTIONS'  && $identifier->getVersion() == $version && $identifier->getPackage() == $package && $identifier->getApiAction() == $apiAction) {
```
Comment (lulco) [inline src/ApiDecider.php]: I think $this->globalPreflight === true is missing in this condition

File: tests/ApiDeciderTest.php [hunk 2]
```diff
@@ -68,4 +68,26 @@ public function testGetHandlers()
...
         $this->assertEquals(1, count($apiDecider->getHandlers()));
     }
+
+    public function testGlobalPreflight()
+    {
+        $linkGenerator = new LinkGenerator(new SimpleRouter([]), new Url('http://test/'));
+        $apiLink = new ApiLink($linkGenerator);
+
+        $apiDecider = new ApiDecider($apiLink);
+        $apiDecider->enableGlobalPreflight();
```
Comment (lulco) [inline tests/ApiDeciderTest.php]: try to run test without this row, result will be the same (but it shouldn't be)

Comment (tomaj) [inline tests/ApiDeciderTest.php]: fixed

File: src/ApiDecider.php (hunk 1 above)
Comment (tomaj) [inline src/ApiDecider.php]: fixed

File: README.md [hunk 3]
```diff
@@ -259,6 +259,34 @@ In Nette-Api if you would like to specify IP restrictions for tokens you can use
...
+```
+
+Or you can register custom OPTIONS endpoints:
+
+``` yaml
+services:
+    apiDecider:
+        class: Tomaj\NetteApi\ApiDecider
+        setup:
+            - addApiHandler(\Tomaj\NetteApi\EndpointIdentifier('OPTIONS', 1, 'users'), \Tomaj\NetteApi\Test\Handler\CorsPreflightHandler(), Tomaj\NetteApi\Authorization\NoAuthorization())
```
Comment (lulco) [inline README.md]: remove "Test" from CorsPreflightHandler namespace

File: tests/Handler/CorsPreflightHandlerTest.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        $defaultHandler = new CorsPreflightHandler(new Response());
+        $this->assertNull($defaultHandler->getEndpoint());
+
+        $endpointIdentifier = new EndpointIdentifier('OPTIONS', 1, 'article', 'detail');
+        $defaultHandler->setEndpointIdentifier($endpointIdentifier);
+        $this->assertEquals($endpointIdentifier, $defaultHandler->getEndpoint());
+    }
+
+    /**
+     * @expectedException \Nette\InvalidStateException
```
Comment (lulco) [inline tests/Handler/CorsPreflightHandlerTest.php]: use

File: tests/Handler/CorsPreflightHandlerTest.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    public function testExceptionWhenCreatingLinkWithoutLinkGenerator()
+    {
+        $defaultHandler = new CorsPreflightHandler(new Response());
+        $defaultHandler->createLink([]);
+    }
+
+    /**
+     * @expectedException \Nette\InvalidStateException
```
Comment (lulco) [inline tests/Handler/CorsPreflightHandlerTest.php]: use

File: README.md [hunk 6]
```diff
@@ -259,6 +259,34 @@ In Nette-Api if you would like to specify IP restrictions for tokens you can use
...
+    apiDecider:
+        class: Tomaj\NetteApi\ApiDecider
+        setup:
+            - enableGlobalPreflight()
+            - addApiHandler(\Tomaj\NetteApi\EndpointIdentifier('GET', 1, 'users'), \App\MyApi\v1\Handlers\UsersListingHandler(), Tomaj\NetteApi\Authorization\NoAuthorization())
+```
+
+Or you can register custom OPTIONS endpoints:
+
+``` yaml
```
Comment (lulco) [inline README.md]: neon

File: tests/Handler/CorsPreflightHandlerTest.php (hunk 4 above)
Comment (tomaj) [inline tests/Handler/CorsPreflightHandlerTest.php]: fixed

File: tests/Handler/CorsPreflightHandlerTest.php (hunk 5 above)
Comment (tomaj) [inline tests/Handler/CorsPreflightHandlerTest.php]: fixed

File: README.md (hunk 6 above)
Comment (tomaj) [inline README.md]: fixed

File: README.md (hunk 3 above)
Comment (tomaj) [inline README.md]: fixed

File: tests/Handler/CorsPreflightHandlerTest.php (hunk 5 above)
Comment (tomaj) [inline tests/Handler/CorsPreflightHandlerTest.php]: a had to remove this use statements, looks like there is problem with phpunit if i use use statement

**Comment by lulco** [comment]
:+1:

**Pull request merged**